File: app/src/main/java/jp/hazuki/yuzubrowser/provider/SuggestProvider.java
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.provider;

import android.app.SearchManager;
import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.Context;
import android.content.UriMatcher;
import android.database.AbstractCursor;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.net.Uri;
import android.provider.BaseColumns;
import android.text.TextUtils;

import com.fasterxml.jackson.core.JsonParser;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import jp.hazuki.yuzubrowser.BuildConfig;
import jp.hazuki.yuzubrowser.ErrorReportServer;
import jp.hazuki.yuzubrowser.core.utility.log.Logger;
import jp.hazuki.yuzubrowser.legacy.search.suggest.ISuggest;
import jp.hazuki.yuzubrowser.legacy.search.suggest.SuggestBing;
import jp.hazuki.yuzubrowser.legacy.search.suggest.SuggestDuckDuckGo;
import jp.hazuki.yuzubrowser.legacy.search.suggest.SuggestGoogle;
import jp.hazuki.yuzubrowser.legacy.search.suggest.Suggestion;
import jp.hazuki.yuzubrowser.legacy.settings.data.AppData;
import jp.hazuki.yuzubrowser.legacy.utils.JsonUtils;

public class SuggestProvider extends ContentProvider {
    private static final String TAG = "GoogleSuggestProvider";
    private static final String AUTHORITY = BuildConfig.APPLICATION_ID + ".search.SuggestProvider";
    public static final Uri URI_NET = Uri.parse("content://" + AUTHORITY + "/net");
    public static final Uri URI_LOCAL = Uri.parse("content://" + AUTHORITY + "/local");
    public static final Uri URI_NORMAL = Uri.parse("content://" + AUTHORITY + "/normal");
    public static final Uri URI_NONE = Uri.EMPTY;
    public static final String SUGGEST_HISTORY = "suggest_history";
    private static final int TYPE_NET_ALL = 1;
    private static final int TYPE_LOCAL_ALL = 2;
    private static final int TYPE_NORMAL_ALL = 3;
    private static final int TYPE_NET = 4;
    private static final int TYPE_LOCAL = 5;
    private static final int TYPE_NORMAL = 6;
    private static final UriMatcher sUriMatcher = new UriMatcher(UriMatcher.NO_MATCH);

    static {
        sUriMatcher.addURI(AUTHORITY, "net", TYPE_NET);
        sUriMatcher.addURI(AUTHORITY, "local", TYPE_LOCAL);
        sUriMatcher.addURI(AUTHORITY, "normal", TYPE_NORMAL);
        sUriMatcher.addURI(AUTHORITY, "net/*", TYPE_NET_ALL);
        sUriMatcher.addURI(AUTHORITY, "local/*", TYPE_LOCAL_ALL);
        sUriMatcher.addURI(AUTHORITY, "normal/*", TYPE_NORMAL_ALL);
    }

    private static final String[] yuzuPrefix = {
            "yuzu:bookmarks", "yuzu:debug", "yuzu:downloads", "yuzu:history", "yuzu:home", "yuzu:readItLater", "yuzu:resBlock", "yuzu:settings", "yuzu:speeddial"
    };

    private static final int COL_ID = 0;
    //private static final int COL_TEXT_1 = 1;
    //private static final int COL_TEXT_2 = 2;
    //private static final int COL_ICON_1 = 3;
    //private static final int COL_ICON_2 = 4;
    private static final int COL_QUERY = 5;
    private static final int COL_HISTORY = 6;

    private static final String[] COLUMNS = new String[]{
            BaseColumns._ID,
            SearchManager.SUGGEST_COLUMN_TEXT_1,
            SearchManager.SUGGEST_COLUMN_TEXT_2,
            SearchManager.SUGGEST_COLUMN_ICON_1,
            SearchManager.SUGGEST_COLUMN_ICON_2,
            SearchManager.SUGGEST_COLUMN_QUERY,
            SUGGEST_HISTORY,
    };

    private static final String DB_NAME = "searchsuggest.db";
    private static final int DB_VERSION = 1;
    private static final String TABLE_NAME = "main_table1";

    private DatabaseHelper mOpenHelper;

    private int mSuggestType;
    private ISuggest mSuggestEngine;

    @Override
    public boolean onCreate() {
        mOpenHelper = new DatabaseHelper(getContext());
        mSuggestType = AppData.search_suggest_engine.get();
        mSuggestEngine = getSuggestEngine(mSuggestType);
        return true;
    }

    @Override
    public Cursor query(@NonNull Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {
        int type = sUriMatcher.match(uri);
        if (type == UriMatcher.NO_MATCH) {
            Logger.e(TAG, "UriMatcher.NO_MATCH");
            return null;
        }

        String query = uri.getQueryParameter("q");
        if (query != null) query = query.trim();
        switch (type) {
            case TYPE_NET_ALL:
            case TYPE_NET:
                return queryNet(query);
            case TYPE_LOCAL_ALL:
            case TYPE_LOCAL:
                return queryLocal(query);
            case TYPE_NORMAL_ALL:
            case TYPE_NORMAL:
                return queryBoth(query);
        }
        return null;
    }

    private Cursor queryBoth(String query) {
        if (TextUtils.isEmpty(query)) {
            return queryLocal(query);
        }

        try {
            List<Suggestion> net = getSuggests(query);
            if (net != null) {
                String dbQuery = query.replace("%", "$%").replace("_", "$_");
                List<Suggestion> suggestions = new ArrayList<>();

                SQLiteDatabase db = mOpenHelper.getReadableDatabase();
                Cursor c = db.query(TABLE_NAME, null, SearchManager.SUGGEST_COLUMN_QUERY + " LIKE '%' || ? || '%' ESCAPE '$'", new String[]{dbQuery}, null, null, BaseColumns._ID + " DESC", "3");
                int COL_QUERY = c.getColumnIndex(SearchManager.SUGGEST_COLUMN_QUERY);
                while (c.moveToNext()) {
                    Suggestion suggestion = new Suggestion(c.getString(COL_QUERY), true);

                    suggestions.add(suggestion);
                    net.remove(suggestion);
                }
                c.close();

                suggestions.addAll(net);

                for (String prefix : yuzuPrefix) {
                    if (prefix.startsWith(query)) {
                        suggestions.add(new Suggestion(prefix));
                    }
                }

                return new SuggestionsCursor(suggestions);
            }
        } catch (UnknownHostException e) {
            return queryLocal(query);
        }

        return null;
    }

    private Cursor queryNet(String query) {
        if (TextUtils.isEmpty(query)) {
            return null;
        }

        try {
            List<Suggestion> list = getSuggests(query);
            if (list != null) {
                for (String prefix : yuzuPrefix) {
                    if (prefix.startsWith(query)) {
                        list.add(new Suggestion(prefix));
                    }
                }
                return new SuggestionsCursor(list);
            }
        } catch (UnknownHostException e) {
            return queryLocal(query);
        }

        return null;
    }

    private List<Suggestion> getSuggests(String query) throws UnknownHostException {
        if (AppData.search_suggest_engine.get() != mSuggestType) {
            mSuggestType = AppData.search_suggest_engine.get();
            mSuggestEngine = getSuggestEngine(mSuggestType);
        }

        JsonParser parser = null;
        try {
            URL url = mSuggestEngine.getUrl(query);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setConnectTimeout(2000);
            connection.setRequestMethod("GET");
            connection.connect();
            if (connection.getResponseCode() == 200) {
                parser = JsonUtils.getFactory().createParser(connection.getInputStream());

                return mSuggestEngine.getSuggestions(parser);
            }
        } catch (UnknownHostException e) {
            throw e;
        } catch (IOException | IllegalStateException | ArrayIndexOutOfBoundsException e) {
            // ArrayIndexOutOfBoundsException - workaround for OkHttp
            ErrorReportServer.printAndWriteLog(e);
        } finally {
            if (parser != null)
                try {
                    parser.close();
                } catch (IOException e) {
                    ErrorReportServer.printAndWriteLog(e);
                }
        }
        return null;
    }

    private Cursor queryLocal(String query) {
        SQLiteDatabase db = mOpenHelper.getReadableDatabase();
        if (TextUtils.isEmpty(query))
            return wrapCursor(db.query(TABLE_NAME, null, null, null, null, null, BaseColumns._ID + " DESC"));
        else {
            String dbQuery = query.replace("%", "$%").replace("_", "$_");
            return addYuzuPrefix(query, db.query(TABLE_NAME, null, SearchManager.SUGGEST_COLUMN_QUERY + " LIKE '%' || ? || '%' ESCAPE '$'", new String[]{dbQuery}, null, null, BaseColumns._ID + " DESC"));
        }
    }

    private Cursor addYuzuPrefix(String query, Cursor c) {
        ArrayList<Suggestion> suggestions = new ArrayList<>();
        if (!TextUtils.isEmpty(query)) {
            for (String prefix : yuzuPrefix) {
                if (prefix.startsWith(query)) {
                    suggestions.add(new Suggestion(prefix));
                }
            }
        }
        if (c != null) {
            final int COL_QUERY = c.getColumnIndex(SearchManager.SUGGEST_COLUMN_QUERY);
            while (c.moveToNext()) {
                suggestions.add(new Suggestion(c.getString(COL_QUERY), true));
            }
            c.close();
        }
        return new SuggestionsCursor(suggestions);
    }

    private Cursor wrapCursor(Cursor c) {
        ArrayList<Suggestion> suggestions = new ArrayList<>();
        if (c != null) {
            final int COL_QUERY = c.getColumnIndex(SearchManager.SUGGEST_COLUMN_QUERY);
            while (c.moveToNext()) {
                suggestions.add(new Suggestion(c.getString(COL_QUERY), true));
            }
            c.close();
        }
        return new SuggestionsCursor(suggestions);
    }

    private ISuggest getSuggestEngine(int type) {
        switch (type) {
            case 1:
                return new SuggestBing();
            case 2:
                return new SuggestDuckDuckGo();
            default:
                return new SuggestGoogle();
        }
    }

    public static class SuggestionsCursor extends AbstractCursor {
        private List<Suggestion> mList;

        public SuggestionsCursor(List<Suggestion> list) {
            mList = list;
        }

        @Override
        public String[] getColumnNames() {
            return COLUMNS;
        }

        @Override
        public int getCount() {
            return mList.size();
        }

        @Override
        public String getString(int column) {
            if (getPosition() == -1) return null;
            switch (column) {
                case COL_ID:
                    return String.valueOf(getPosition());
                //case COL_TEXT_1:
                case COL_QUERY:
                    return mList.get(getPosition()).getTitle();
            }
            return null;
        }

        @Override
        public long getLong(int column) {
            if (column == COL_ID) {
                return getPosition();
            }
            throw new UnsupportedOperationException();
        }

        @Override
        public double getDouble(int column) {
            throw new UnsupportedOperationException();
        }

        @Override
        public float getFloat(int column) {
            throw new UnsupportedOperationException();
        }

        @Override
        public int getInt(int column) {
            if (getPosition() == -1) return 0;
            if (column == COL_HISTORY) {
                return mList.get(getPosition()).getSuggestHistory() ? 1 : 0;
            } else {
                return 0;
            }
        }

        @Override
        public short getShort(int column) {
            throw new UnsupportedOperationException();
        }

        @Override
        public boolean isNull(int column) {
            throw new UnsupportedOperationException();
        }
    }

    @Override
    public int delete(@NonNull Uri uri, String selection, String[] selectionArgs) {
        int type = sUriMatcher.match(uri);
        if (type == TYPE_NET || type == TYPE_NET_ALL)
            return 0;
        SQLiteDatabase db = mOpenHelper.getWritableDatabase();
        return db.delete(TABLE_NAME, selection, selectionArgs);
    }

    @Override
    public String getType(@NonNull Uri uri) {
        return SearchManager.SUGGEST_MIME_TYPE;
    }

    @Override
    public Uri insert(@NonNull Uri uri, ContentValues values) {
        int type = sUriMatcher.match(uri);
        if (type == TYPE_NET || type == TYPE_NET_ALL)
            return null;
        SQLiteDatabase db = mOpenHelper.getWritableDatabase();
        db.insert(TABLE_NAME, null, values);
        return null;
    }

    @Override
    public int update(@NonNull Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        throw new UnsupportedOperationException();
    }

    private static class DatabaseHelper extends SQLiteOpenHelper {
        DatabaseHelper(Context context) {
            super(context, DB_NAME, null, DB_VERSION);
        }

        @Override
        public void onCreate(SQLiteDatabase db) {
            db.beginTransaction();
            try {
                db.execSQL("CREATE TABLE " + TABLE_NAME + " (" +
                        BaseColumns._ID + " INTEGER PRIMARY KEY" +
                        ", " + SearchManager.SUGGEST_COLUMN_QUERY + " TEXT UNIQUE ON CONFLICT REPLACE" +
                        ")");
                db.setTransactionSuccessful();
            } finally {
                db.endTransaction();
            }
        }

        @Override
        public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
            db.execSQL("DROP TABLE IF EXISTS " + TABLE_NAME);
            onCreate(db);
        }
    }
}

File: app/src/main/java/jp/hazuki/yuzubrowser/provider/SuggestProvider.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.provider

import android.app.SearchManager
import android.content.ContentValues
import android.content.Context
import android.content.UriMatcher
import android.database.AbstractCursor
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.net.Uri
import android.provider.BaseColumns
import android.text.TextUtils
import com.squareup.moshi.JsonDataException
import com.squareup.moshi.JsonReader
import dagger.android.DaggerContentProvider
import jp.hazuki.yuzubrowser.BuildConfig
import jp.hazuki.yuzubrowser.ErrorReportServer
import jp.hazuki.yuzubrowser.core.utility.log.Logger
import jp.hazuki.yuzubrowser.search.model.SearchSuggestModel
import jp.hazuki.yuzubrowser.search.model.suggest.ISuggest
import jp.hazuki.yuzubrowser.search.model.suggest.SuggestBing
import jp.hazuki.yuzubrowser.search.model.suggest.SuggestDuckDuckGo
import jp.hazuki.yuzubrowser.search.model.suggest.SuggestGoogle
import jp.hazuki.yuzubrowser.ui.settings.AppPrefs
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.net.UnknownHostException
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class SuggestProvider : DaggerContentProvider() {

    private lateinit var mOpenHelper: DatabaseHelper
    private lateinit var mSuggestEngine: ISuggest
    @Inject
    lateinit var okHttpClient: OkHttpClient

    private var mSuggestType: Int = 0

    override fun onCreate(): Boolean {
        super.onCreate()
        val context = context ?: throw IllegalStateException()
        mOpenHelper = DatabaseHelper(context)
        mSuggestType = AppPrefs.search_suggest_engine.get()
        mSuggestEngine = getSuggestEngine(mSuggestType)
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        val type = sUriMatcher.match(uri)
        if (type == UriMatcher.NO_MATCH) {
            Logger.e(TAG, "UriMatcher.NO_MATCH")
            return null
        }

        var query = uri.getQueryParameter("q")
        if (query != null) query = query.trim { it <= ' ' }
        when (type) {
            TYPE_NET_ALL, TYPE_NET -> return queryNet(query)
            TYPE_LOCAL_ALL, TYPE_LOCAL -> return queryLocal(query)
            TYPE_NORMAL_ALL, TYPE_NORMAL -> return queryBoth(query)
        }
        return null
    }

    private fun queryBoth(query: String?): Cursor? {
        if (query.isNullOrEmpty()) {
            return queryLocal(query)
        }

        try {
            val net = getSuggests(query)
            if (net != null) {
                val dbQuery = query.replace("%", "$%").replace("_", "\$_")
                val suggestions = ArrayList<SearchSuggestModel.SuggestModel>()

                val db = mOpenHelper.readableDatabase
                val c = db.query(TABLE_NAME, null, SearchManager.SUGGEST_COLUMN_QUERY + " LIKE '%' || ? || '%' ESCAPE '$'", arrayOf(dbQuery), null, null, BaseColumns._ID + " DESC", "3")
                val colQuery = c.getColumnIndex(SearchManager.SUGGEST_COLUMN_QUERY)
                while (c.moveToNext()) {
                    val suggestion = SearchSuggestModel.SuggestModel(c.getString(colQuery), true)

                    suggestions.add(suggestion)
                    net.remove(suggestion)
                }
                c.close()

                suggestions.addAll(net)

                for (prefix in yuzuPrefix) {
                    if (prefix.startsWith(query)) {
                        suggestions.add(SearchSuggestModel.SuggestModel(prefix))
                    }
                }

                return SuggestionsCursor(suggestions)
            }
        } catch (e: UnknownHostException) {
            return queryLocal(query)
        }

        return null
    }

    private fun queryNet(query: String?): Cursor? {
        if (query.isNullOrEmpty()) {
            return null
        }

        try {
            val list = getSuggests(query)
            if (list != null) {
                for (prefix in yuzuPrefix) {
                    if (prefix.startsWith(query)) {
                        list.add(SearchSuggestModel.SuggestModel(prefix))
                    }
                }
                return SuggestionsCursor(list)
            }
        } catch (e: UnknownHostException) {
            return queryLocal(query)
        }

        return null
    }

    @Throws(UnknownHostException::class)
    private fun getSuggests(query: String): MutableList<SearchSuggestModel.SuggestModel>? {
        if (AppPrefs.search_suggest_engine.get() != mSuggestType) {
            mSuggestType = AppPrefs.search_suggest_engine.get()
            mSuggestEngine = getSuggestEngine(mSuggestType)
        }

        val url = mSuggestEngine.getUrl(query)
        val client = okHttpClient.newBuilder()
                .connectTimeout(2000, TimeUnit.MILLISECONDS)
                .build()
        val request = Request.Builder()
                .url(url)
                .get()
                .build()

        val call = client.newCall(request)
        try {
            val response = call.execute()
            if (response.isSuccessful) {
                val body = response.body
                if (body != null) {
                    JsonReader.of(body.source()).use {
                        return mSuggestEngine.getSuggestions(it)
                    }
                }
            }
        } catch (e: Exception) {
            when (e) {
                is IOException, is JsonDataException -> {
                    ErrorReportServer.printAndWriteLog(e)
                }
                else -> throw e
            }
        }
        return null
    }

    private fun queryLocal(query: String?): Cursor {
        val db = mOpenHelper.readableDatabase
        return if (TextUtils.isEmpty(query))
            wrapCursor(db.query(TABLE_NAME, null, null, null, null, null, BaseColumns._ID + " DESC"))
        else {
            val dbQuery = query!!.replace("%", "$%").replace("_", "\$_")
            addYuzuPrefix(query, db.query(TABLE_NAME, null, SearchManager.SUGGEST_COLUMN_QUERY + " LIKE '%' || ? || '%' ESCAPE '$'", arrayOf(dbQuery), null, null, BaseColumns._ID + " DESC"))
        }
    }

    private fun addYuzuPrefix(query: String, c: Cursor?): Cursor {
        val suggestions = ArrayList<SearchSuggestModel.SuggestModel>()
        if (!TextUtils.isEmpty(query)) {
            for (prefix in yuzuPrefix) {
                if (prefix.startsWith(query)) {
                    suggestions.add(SearchSuggestModel.SuggestModel(prefix))
                }
            }
        }
        if (c != null) {
            val colQuery = c.getColumnIndex(SearchManager.SUGGEST_COLUMN_QUERY)
            while (c.moveToNext()) {
                suggestions.add(SearchSuggestModel.SuggestModel(c.getString(colQuery), true))
            }
            c.close()
        }
        return SuggestionsCursor(suggestions)
    }

    private fun wrapCursor(c: Cursor?): Cursor {
        val suggestions = ArrayList<SearchSuggestModel.SuggestModel>()
        if (c != null) {
            val colQuery = c.getColumnIndex(SearchManager.SUGGEST_COLUMN_QUERY)
            while (c.moveToNext()) {
                suggestions.add(SearchSuggestModel.SuggestModel(c.getString(colQuery), true))
            }
            c.close()
        }
        return SuggestionsCursor(suggestions)
    }

    private fun getSuggestEngine(type: Int): ISuggest {
        return when (type) {
            1 -> SuggestBing()
            2 -> SuggestDuckDuckGo()
            else -> SuggestGoogle()
        }
    }

    class SuggestionsCursor(private val mList: List<SearchSuggestModel.SuggestModel>) : AbstractCursor() {

        override fun getColumnNames(): Array<String> {
            return COLUMNS
        }

        override fun getCount(): Int {
            return mList.size
        }

        override fun getString(column: Int): String? {
            if (position == -1) return null
            when (column) {
                COL_ID -> return position.toString()
                //case COL_TEXT_1:
                COL_QUERY -> return mList[position].suggest
            }
            return null
        }

        override fun getLong(column: Int): Long {
            if (column == COL_ID) {
                return position.toLong()
            }
            throw UnsupportedOperationException()
        }

        override fun getDouble(column: Int): Double {
            throw UnsupportedOperationException()
        }

        override fun getFloat(column: Int): Float {
            throw UnsupportedOperationException()
        }

        override fun getInt(column: Int): Int {
            if (position == -1) return 0
            return if (column == COL_HISTORY) {
                if (mList[position].suggestHistory) 1 else 0
            } else {
                0
            }
        }

        override fun getShort(column: Int): Short {
            throw UnsupportedOperationException()
        }

        override fun isNull(column: Int): Boolean {
            throw UnsupportedOperationException()
        }
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        val type = sUriMatcher.match(uri)
        if (type == TYPE_NET || type == TYPE_NET_ALL)
            return 0
        val db = mOpenHelper.writableDatabase
        return db.delete(TABLE_NAME, selection, selectionArgs)
    }

    override fun getType(uri: Uri): String? {
        return SearchManager.SUGGEST_MIME_TYPE
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        val type = sUriMatcher.match(uri)
        if (type == TYPE_NET || type == TYPE_NET_ALL)
            return null
        val db = mOpenHelper.writableDatabase
        db.insert(TABLE_NAME, null, values)
        return null
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {
        throw UnsupportedOperationException()
    }

    private class DatabaseHelper internal constructor(context: Context) : SQLiteOpenHelper(context, DB_NAME, null, DB_VERSION) {

        override fun onCreate(db: SQLiteDatabase) {
            db.beginTransaction()
            try {
                db.execSQL("CREATE TABLE " + TABLE_NAME + " (" +
                        BaseColumns._ID + " INTEGER PRIMARY KEY" +
                        ", " + SearchManager.SUGGEST_COLUMN_QUERY + " TEXT UNIQUE ON CONFLICT REPLACE" +
                        ")")
                db.setTransactionSuccessful()
            } finally {
                db.endTransaction()
            }
        }

        override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
            db.execSQL("DROP TABLE IF EXISTS $TABLE_NAME")
            onCreate(db)
        }
    }

    companion object {
        private const val TAG = "GoogleSuggestProvider"
        private const val AUTHORITY = BuildConfig.APPLICATION_ID + ".search.SuggestProvider"
        val URI_NET: Uri = Uri.parse("content://$AUTHORITY/net")
        val URI_LOCAL: Uri = Uri.parse("content://$AUTHORITY/local")
        val URI_NORMAL: Uri = Uri.parse("content://$AUTHORITY/normal")
        val URI_NONE: Uri = Uri.EMPTY
        const val SUGGEST_HISTORY = "suggest_history"
        private const val TYPE_NET_ALL = 1
        private const val TYPE_LOCAL_ALL = 2
        private const val TYPE_NORMAL_ALL = 3
        private const val TYPE_NET = 4
        private const val TYPE_LOCAL = 5
        private const val TYPE_NORMAL = 6
        private val sUriMatcher = UriMatcher(UriMatcher.NO_MATCH)

        init {
            sUriMatcher.addURI(AUTHORITY, "net", TYPE_NET)
            sUriMatcher.addURI(AUTHORITY, "local", TYPE_LOCAL)
            sUriMatcher.addURI(AUTHORITY, "normal", TYPE_NORMAL)
            sUriMatcher.addURI(AUTHORITY, "net/*", TYPE_NET_ALL)
            sUriMatcher.addURI(AUTHORITY, "local/*", TYPE_LOCAL_ALL)
            sUriMatcher.addURI(AUTHORITY, "normal/*", TYPE_NORMAL_ALL)
        }

        private val yuzuPrefix = arrayOf("yuzu:bookmarks", "yuzu:debug", "yuzu:downloads", "yuzu:history", "yuzu:home", "yuzu:readItLater", "yuzu:resBlock", "yuzu:settings", "yuzu:speeddial")

        private const val COL_ID = 0
        //private static final int COL_TEXT_1 = 1;
        //private static final int COL_TEXT_2 = 2;
        //private static final int COL_ICON_1 = 3;
        //private static final int COL_ICON_2 = 4;
        private const val COL_QUERY = 5
        private const val COL_HISTORY = 6

        private val COLUMNS = arrayOf(BaseColumns._ID, SearchManager.SUGGEST_COLUMN_TEXT_1, SearchManager.SUGGEST_COLUMN_TEXT_2, SearchManager.SUGGEST_COLUMN_ICON_1, SearchManager.SUGGEST_COLUMN_ICON_2, SearchManager.SUGGEST_COLUMN_QUERY, SUGGEST_HISTORY)

        private const val DB_NAME = "searchsuggest.db"
        private const val DB_VERSION = 1
        private const val TABLE_NAME = "main_table1"
    }
}
